### ISSUE TITLE
Domain inclusion seem to not work

### ISSUE THREAD

**Issue by cs-shadow** [cs-shadow/buildstream#82]
In GitLab by @bochecha on Aug 27, 2017, 10:20

I've got the following in my `project.conf`: 

```yaml
split-rules:
  devel:
  # Those are the default BuildStream ones
  - |
    %{includedir}
  - |
    %{includedir}/**
  - |
    %{libdir}/lib*.a
  - |
    %{libdir}/lib*.la
  - |
    %{libdir}/pkgconfig/*.pc
  - |
    %{datadir}/pkgconfig/*.pc
  - |
    %{datadir}/aclocal/*.m4
  # Those are the ones I added
  - |
    %{bindir}/*gcc*
  - |
    %{libexecdir}/gcc
```

I have a `compose` element with:

```yaml
config:
  include:
  - runtime
  - locale
```

I would expect that element not to have any of the things I added to the `devel` domain. And yet, when I `bst checkout` it, there they are.

Am I doing something wrong? Or did I stumble upon a bug?

(edited to explicitly mention I was using a `compose` element)

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Aug 27, 2017, 17:01

closed

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Aug 27, 2017, 17:01

Yeah it doesnt work that way (we should really fix #24 soon, that would have saved you some time...).

[Split rules](https://buildstream.gitlab.io/buildstream/public.html#split-rules) are used to define the categorization, but the only ways to leverage that currently is with:
* [Compose elements](https://buildstream.gitlab.io/buildstream/elements/compose.html#module-elements.compose) (which is the element that exposes the `include:` semantic you are trying to use)
* The [Element.stage_artifact()](https://buildstream.gitlab.io/buildstream/buildstream.element.html#buildstream.element.Element.stage_artifact) API, which can be used by plugins in similar ways that the `compose` element does.

It is expected to be useful to any elements that are deployment related (e.g. I think the dpkg deployment that @jonathanmaw wrote uses the same internal artifact staging API and uses split rules to specify what goes into a package).

**Comment by cs-shadow** [comment]
In GitLab by @bochecha on Aug 27, 2017, 18:32

> [Compose elements](https://buildstream.gitlab.io/buildstream/elements/compose.html#module-elements.compose) (which is the element that exposes the `include:` semantic you are trying to use)

I **am** using a `compose` element, that's where I put the `include` rules. :confused:

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Aug 27, 2017, 19:52

reopened

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Aug 27, 2017, 19:53

Aha, can you show me the compose element ?

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Aug 27, 2017, 20:41

Another thing comes to mind, I'm not sure where you defined the split rules (in your `project.conf` ?).

You might check the output of `bst show --deps none --format %{public} element.bst` for an element going into the composition and see if it's what you expect (the *split-rules* for the compose element itself is meaningless for the output of a compose element; it's the *split-rules* found on the elements going into the composition which matter).

**Comment by cs-shadow** [comment]
In GitLab by @bochecha on Aug 28, 2017, 07:28

changed the description

**Comment by cs-shadow** [comment]
In GitLab by @bochecha on Aug 28, 2017, 10:59

> Aha, can you show me the compose element ?

Sure, here it is: https://framagit.org/bochecha/org.ideascube.Sdk/blob/master/elements/org.ideascube.Platform.bst

> Another thing comes to mind, I'm not sure where you defined the split rules (in your `project.conf` ?).

Yup, it's here: https://framagit.org/bochecha/org.ideascube.Sdk/blob/master/project.conf

> the _split-rules_ for the compose element itself is meaningless for the output of a compose element; it's the _split-rules_ found on the elements going into the composition which matter

I'm not sure I understand what you mean here. :confused:

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Aug 28, 2017, 20:56

Looks like you've spotted a good problem here...

> > the split-rules for the compose element itself is meaningless for the output of a compose element; it's the split-rules found on the elements going into the composition which matter

> I'm not sure I understand what you mean here. :confused:

A bit tangential to the issue but *split-rules* defined in the project.conf are defaults apply to all project elements while those can be overridden individually on a per element basis in the `public:` data.


Aaanyway, I figured out what is going on...


What you have in your `project.conf` is something that defines `/usr/bin/gcc` to be in the devel category. However, `/usr/bin/gcc` is already by default in the `runtime` domain which is being included; so it is still claimed and shipped.

Surely it does not make sense to create a runtime domain while hand specifying each and every binary in `/usr/bin` that you want, so overriding the default `runtime` domain is not viable.


So, now it seems we need some more flexibility; it could be achieved in a few different ways and I wonder which makes sense and is easy to understand and use.

### Regex

We used to have regexes which I thought to be not user friendly, these could allow one to express an *everything except* pattern to use when overriding the `runtime` domain.

Not a good UX though, you still need to override the default domain instead of just specifying one pattern match.

### Exclusion rules

This could be implemented either as a feature of the `compose` element, or an inherent structure of the public *split-rules* data, and each would effect the user experience in different ways.

#### Compose excluding domains

Currently the compose element has `include:` to define what to include, we could add `exclude:` and define it to mean that we exclude files from this domain even if they also appear in included domains.

This seems to be simple all around; while it would handle this case well, I'm not sure if it's sufficiently flexible.

#### Public data exclude domains

Instead of defining lists of files to include for each public data domain; also define lists of files to exclude for each public data domain. This keeps inclusion and exclusion definitions closely coupled with each domain (and different exclusions might apply for different domains).

This would make the override approach possibly a bit lighter; because you could probably get away with just overriding the exclude list, still it seems more typing than using the `compose` element `exclude:` approach.

Also uncertain whether this would be flexible enough.


Any thoughts ?

**Comment by cs-shadow** [comment]
In GitLab by @bochecha on Aug 28, 2017, 21:03

I like the idea of writing `exclude:` in my `compose` element, rather than `include:`-ing.

In fact, I would find this much more natural, but maybe I'm biased by `flatpak-builder`. :smile: 

You need the build tools in order to build things. Therefore it feels much more natural to build an sdk, and then remove from it the unneeded stuff to make a runtime.

I don't know whether that would be sufficiently flexible for other use-cases. But then, what other use-cases are there, really?

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Aug 28, 2017, 21:15

> I don't know whether that would be sufficiently flexible for other use-cases. But then, what other use-cases are there, really?

Many can be rather similar; for instance one thing I want to do soon is output docker images.

When outputting bootable images or `/usr/` filesystem images for atomic updates on an *ostree as host* systems; also the use cases are fairly straight forward and similar.

Other than those, we have debian packaging elements which by default will use the *split-rules* to identify what gets put into a package; they will expect some debian tooling to be available in the runtime and will stage an individual build output and create packages with it (there is another case of using rpm but I haven't seen patches for that yet).


In any case I think I like enhancing the `compose` element for this more than anything else; this way we keep the public *split-rules* data very simple.

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Aug 29, 2017, 18:13

mentioned in commit b2584a108c5baafa0bd09d6d608d6f1d23534280

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Aug 29, 2017, 20:03

mentioned in commit 3cc9e191b83526cea757f92b522e3e8b1359d6eb

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Aug 29, 2017, 22:05

closed via commit 3cc9e191b83526cea757f92b522e3e8b1359d6eb

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Aug 29, 2017, 22:18

Hmmm I thought I had disabled auto-close of bugs from commit messages...

Anyway now the `compose` element has an `exclude` semantic and the integration tests are updated with a new test for that, documented [here](https://buildstream.gitlab.io/buildstream/elements/compose.html#module-elements.compose)
